### DISCUSSION TITLE
Support for Ultra HDR

### DISCUSSION THREAD

**Discussion by cromefire** [libvips/libvips#3798] [Ideas]
[Ultra HDR](https://developer.android.com/guide/topics/media/platform/hdr-image-format) is Google's new HDR image format based on JPEG and GainMaps (for compatibility) and newer Pixels actually produce Ultra HDR photos by default now (it's pretty cool!)

But the problem is right now libvips is running into that mentioned compatibility mode, where it's taking the primary image and discarding the GainMap, so any image ever processed by libvips looses its HDR and becomes SDR only (it does copy the metadata already so it's identified as Ultra HDR, but it doesn't work).

Here's some random Ultra HDR image I use for testing (it's a screenshot originally at up to 1000nits, so the HDR isn't limited by the camera): ![test ultra](https://github.com/libvips/libvips/assets/26320625/60188b1e-c186-47f9-a536-750c93560907)
Here it is copied (`vips copy`): ![test-copy ultra](https://github.com/libvips/libvips/assets/26320625/a1a3a212-67b2-41b9-aa71-757263d47ba9)
Here it is flipped (`vips flip`): ![test-flip ultra](https://github.com/libvips/libvips/assets/26320625/5222bc65-f596-4120-8e77-d0a3b1e8bccd)

~~I hope GitHub doesn't modify them~~ (nope looks good on my phone), because in that case the first one should be HDR on an HDR display using Chrome (Desktop and Android) and then you can just download and inspect them, otherwise I'll just upload them as a zip. And that's the exact problem I'm basically facing. Because as soon as Nextcloud for example (which uses Imaginary, which uses bimg, which uses libvips) touches them, it immediately looses the HDR layer.

That would be the first step, which should probably not require any API changes, the second one (which I don't have any use for right now as I don't build directly on this library but could be pretty useful in the future) would probably be to support Ultra HDR conversion to other formats directly using [libultrahdr](https://github.com/google/libultrahdr) using that you could conceivably convert an Ultra HDR image to an HDR AVIF, JXL or HEIC (or JXR for Windows because they don't seem to support anything else...) image.

**Comment by jcupitt** [comment]
Hi @cromefire,

Sure, this sounds interesting. It looks like support for the extra gain map image could be added pretty easily to the libvips JPEG reader. We could just add an image-valued metadata item called `gain-map` (for example) containing the extra information. Generating a float HDR image could then be done in python or whatever using the standard libvips tools.

Let's make this into an enhancement issue.

**Comment by gregbenz** [comment]
Initial support for HDR JPG with gain maps in libvips: https://github.com/libvips/libvips/pull/4645
This includes support for JPG gain maps (a future update to libultrahdr will expand support to AVIF / HEIF)
